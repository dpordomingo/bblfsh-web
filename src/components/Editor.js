import React, { Component } from 'react';
import styled, { injectGlobal } from 'styled-components';
import { Controlled as CodeMirror } from 'react-codemirror2';
import { background } from '../styling/variables';
import { font, border } from '../styling/variables';
import { connect } from 'react-redux';
import { getLanguageMode } from 'state/languages';
import { change as changeCode, selectNodeByPos } from 'state/code';

import 'codemirror/lib/codemirror.css';
import 'codemirror/theme/solarized.css';
import 'codemirror/mode/python/python';
import 'codemirror/mode/clike/clike';

// eslint-disable-next-line
injectGlobal`
  .ReactCodeMirror, .CodeMirror {
    height: 100%;
  }

  .CodeMirror {
    font-family: ${font.family.code};
    font-size: ${font.size.large};
  }
`;

const Container = styled.div`
  height: 100%;
`;

const BookMark = styled.div`
  display: inline-block;
  position: relative;
  width: 0;
  height: 0;

  &::before {
    content: '';
    position: absolute;
    bottom: -5px;
    left: -3px;
    width: 3px;
    height: 19px;
    border: 1px solid ${border.accent};
    border-right-width: 0 !important;
    border-left-width: 2px !important;
  }
`;

export class Editor extends Component {
  get document() {
    return this.editor.getDoc();
  }

  get editor() {
    return this.codemirror.editor;
  }

  selectCode(range) {
    if (this.mark) {
      this.mark.clear();
    }

    if (!range) {
      this.mark = null;
      return;
    }

    const { from, to } = range;

    if (from && to) {
      this.mark = this.document.markText(from, to, {
        css: 'background: ' + background.highlight,
      });
    } else if (from) {
      this.mark = this.document.setBookmark(from, { widget: this.bookmark });
    }
  }

  componentDidUpdate(prevProps) {
    if (prevProps.markRange !== this.props.markRange) {
      this.selectCode(this.props.markRange);
    }
  }

  onCursor(pos) {
    if (!this.props.onCursorChanged) {
      return;
    }

    this.props.onCursorChanged(pos);
  }

  render() {
    const { code, languageMode, onChange } = this.props;

    const options = {
      mode: languageMode,
      lineNumbers: true,
      theme: 'solarized light',
    };

    return (
      <Container>
        <CodeMirror
          ref={r => (this.codemirror = r)}
          autoFocus={true}
          value={code}
          options={options}
          onBeforeChange={(editor, data, v) => onChange(v)}
          onCursor={(_, pos) => this.onCursor(pos)}
        />
        <BookMark
          innerRef={elem => {
            this.bookmark = elem;
          }}
        />
      </Container>
    );
  }
}

export const mapStateToProps = state => {
  return {
    code: state.code.code,
    languageMode: getLanguageMode(state),
    markRange: state.code.markRange,
  };
};

const mapDispatchToProps = {
  onChange: changeCode,
  onCursorChanged: selectNodeByPos,
};

export default connect(mapStateToProps, mapDispatchToProps)(Editor);
