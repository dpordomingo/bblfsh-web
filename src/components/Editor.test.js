import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';
import { Editor, mapStateToProps } from './Editor';

const mkPos = (line, ch) => ({ line, ch });

describe('Editor', () => {
  it('renders without crashing', () => {
    const wrapper = shallow(
      <Editor code="foo = 1" languageMode="python" markRange={null} />
    );

    expect(toJson(wrapper)).toMatchStyledComponentsSnapshot();
  });

  it('calls onCursorChanged when the selection changes', () => {
    const spy = jest.fn();
    const wrapper = mount(
      <Editor code="foo = 1" languageMode="python" onCursorChanged={spy} />
    );

    wrapper.instance().document.setSelection(mkPos(0, 4), mkPos(0, 2));

    // NOTE: there is an initial call when the CodeMirror is rendered.
    expect(spy.mock.calls.length).toBe(2);
    const [pos] = spy.mock.calls[1];
    expect(pos).toEqual(mkPos(0, 2));
  });

  it('calls onCursorChanged when the cursor position changes', () => {
    const spy = jest.fn();
    const wrapper = mount(
      <Editor code="foo = 1" languageMode="python" onCursorChanged={spy} />
    );

    wrapper.instance().document.setCursor(mkPos(0, 4));

    // NOTE: there is an initial call when the CodeMirror is rendered.
    expect(spy.mock.calls.length).toBe(2);
    const [pos] = spy.mock.calls[1];
    expect(pos).toEqual(mkPos(0, 4));
  });

  it('calls onChange when the content changes', () => {
    const spy = jest.fn();
    const wrapper = mount(
      <Editor code="foo = 1" languageMode="python" onChange={spy} />
    );

    wrapper
      .instance()
      .document.replaceRange('changed', mkPos(0, 0), mkPos(0, 3));

    expect(spy.mock.calls.length).toBe(1);
    expect(spy.mock.calls[0][0]).toBe('changed = 1');
  });

  it('selects the code when markRange props is presented', () => {
    const wrapper = mount(
      <Editor
        code="foo = 1"
        languageMode="python"
        markRange={{ from: mkPos(0, 0), to: mkPos(0, 4) }}
      />
    );

    expect(wrapper.instance().document.getAllMarks().length).toBe(1);
  });

  it('change selects the code when markRange is changed', () => {
    const wrapper = mount(
      <Editor
        code="foo = 1"
        languageMode="python"
        markRange={{ from: mkPos(0, 0), to: mkPos(0, 4) }}
      />
    );

    const assertMarks = n =>
      expect(wrapper.instance().document.getAllMarks().length).toBe(n);

    wrapper.setProps({ markRange: null });
    assertMarks(0);

    wrapper.setProps({ markRange: { from: mkPos(0, 0), to: mkPos(0, 4) } });
    assertMarks(1);
  });

  it('mapStateToProps', () => {
    const state = {
      code: {
        code: 'foo = 1',
        markRange: { from: 1, to: 2 },
      },
      languages: {
        languages: {
          python: {
            mode: 'x-python',
          },
        },
        selected: 'python',
      },
    };

    const props = mapStateToProps(state);
    expect(props).toEqual({
      code: 'foo = 1',
      languageMode: 'x-python',
      markRange: { from: 1, to: 2 },
    });
  });
});
