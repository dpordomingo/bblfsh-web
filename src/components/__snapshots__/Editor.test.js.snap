// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor renders without crashing 1`] = `
<styled.div>
  <Controlled
    autoFocus={true}
    onBeforeChange={[Function]}
    onCursor={[Function]}
    options={
      Object {
        "lineNumbers": true,
        "mode": "python",
        "theme": "solarized light",
      }
    }
    value="foo = 1"
  />
  <styled.div
    innerRef={[Function]}
  />
</styled.div>
`;
